File: app/src/main/java/cz/dvratil/fbeventsync/Graph.java
/*
    Copyright (C) 2017  Daniel Vrátil <me@dvratil.cz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package cz.dvratil.fbeventsync;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import com.loopj.android.http.*;

public class Graph {

    public static class AppUsage {

        private static final String PREFS_CALLCOUNT = "appusage_callCount";
        private static final String PREFS_TOTALTIME = "appusage_totalTime";
        private static final String PREFS_TOTALCPUTIME = "appusage_totalCPUTime";

        public int callCount = 0;
        public int totalTime = 0;
        public int totalCPUTime = 0;

        AppUsage(int callCount, int totalTime, int totalCPUTime) {
            this.callCount = callCount;
            this.totalTime = totalTime;
            this.totalCPUTime = totalCPUTime;
        }

        public void store(Context context) {
            SharedPreferences prefs = context.getSharedPreferences(
                    context.getString(R.string.cz_dvratil_fbeventsync_preferences),
                    Context.MODE_PRIVATE);
            SharedPreferences.Editor edit = prefs.edit();
            edit.putInt(PREFS_CALLCOUNT, this.callCount);
            edit.putInt(PREFS_TOTALTIME, this.totalTime);
            edit.putInt(PREFS_TOTALCPUTIME, this.totalCPUTime);
            edit.apply();
        }

        static AppUsage load(Context context) {
            SharedPreferences prefs = context.getSharedPreferences(
                    context.getString(R.string.cz_dvratil_fbeventsync_preferences),
                    Context.MODE_PRIVATE);
            return new AppUsage(
                    prefs.getInt(PREFS_CALLCOUNT, 0),
                    prefs.getInt(PREFS_TOTALTIME, 0),
                    prefs.getInt(PREFS_TOTALCPUTIME, 0));
        }
    }

    private static final String BASE_URL = "https://graph.facebook.com/v2.9/";

    public static final String FIELDS_PARAM = "fields";
    public static final String LIMIT_PARAM = "limit";
    public static final String TYPE_PARAM = "type";
    public static final String AFTER_PARAM = "after";

    public static RequestHandle me(String accessToken, AsyncHttpResponseHandler handler) {
        RequestParams params = new RequestParams();
        params.add("access_token", accessToken);
        params.add(FIELDS_PARAM, "name");
        return new AsyncHttpClient().get(BASE_URL+ "/me", params, handler);
    }

    public static RequestHandle events(String accessToken, RequestParams params, AsyncHttpResponseHandler handler) {
        SyncHttpClient client = new SyncHttpClient();
        params.add("access_token", accessToken);
        return client.get(BASE_URL + "/me/events", params, handler);
    }

    public static RequestHandle refreshTokens(Context context, String scopes, AsyncHttpResponseHandler handler) {
        SyncHttpClient client = new SyncHttpClient();
        RequestParams params = new RequestParams();
        params.put("client_id", context.getString(R.string.facebook_app_id));
        params.put("redirect_uri", "https://www.facebook.com/connect/login_success.html");
        params.put("response_type", "token");
        params.put("scopes", scopes);
        return client.get("https://www.facebook.com/v2.9/dialog/oauth", params, handler);
    }

    public static RequestHandle fetchBirthdayICal(String birthdayICalUri, AsyncHttpResponseHandler handler) {
        SyncHttpClient client = new SyncHttpClient();
        // Pretend we are cURL, so that Facebook does not redirect us to facebook.com/unsupportedbrowser
        client.setUserAgent("curl/7.55.1");
        Log.d("GRAPH", "fetchBirthdayICal: " + birthdayICalUri);
        return client.get(birthdayICalUri, handler);
    }
}

File: app/src/main/java/cz/dvratil/fbeventsync/Graph.kt
/*
    Copyright (C) 2017 - 2018  Daniel Vrátil <me@dvratil.cz>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package cz.dvratil.fbeventsync

import android.content.Context

import com.loopj.android.http.AsyncHttpClient
import com.loopj.android.http.AsyncHttpResponseHandler
import com.loopj.android.http.RequestHandle
import com.loopj.android.http.RequestParams
import com.loopj.android.http.SyncHttpClient

object Graph {

    private const val BASE_URL = "https://graph.facebook.com/v2.9/"

    const val FIELDS_PARAM = "fields"
    const val LIMIT_PARAM = "limit"
    const val AFTER_PARAM = "after"
    const val ACCESS_TOKEN_PARAM = "access_token"

    fun me(accessToken: String, handler: AsyncHttpResponseHandler): RequestHandle {
        val params = RequestParams().apply {
            add(ACCESS_TOKEN_PARAM, accessToken)
            add(FIELDS_PARAM, "name")
        }
        return AsyncHttpClient().get(BASE_URL + "/me", params, handler)
    }

    fun mePicture(accessToken: String, handler: AsyncHttpResponseHandler): RequestHandle {
        var params = RequestParams().apply {
            add(ACCESS_TOKEN_PARAM, accessToken)
        }
        return AsyncHttpClient().get(BASE_URL + "/me/picture", params, handler)
    }

    fun events(accessToken: String, params: RequestParams, handler: AsyncHttpResponseHandler): RequestHandle {
        val client = SyncHttpClient()
        params.add(ACCESS_TOKEN_PARAM, accessToken)
        return client.get(BASE_URL + "/me/events", params, handler)
    }

    fun refreshTokens(context: Context, scopes: String, handler: AsyncHttpResponseHandler): RequestHandle {
        val client = SyncHttpClient()
        val params = RequestParams().apply {
            put("client_id", context.getString(R.string.facebook_app_id))
            put("redirect_uri", "https://www.facebook.com/connect/login_success.html")
            put("response_type", "token")
            put("scopes", scopes)
        }
        return client.get("https://www.facebook.com/v2.9/dialog/oauth", params, handler)
    }

    fun fetchBirthdayICal(birthdayICalUri: String, handler: AsyncHttpResponseHandler): RequestHandle {
        val client = SyncHttpClient()
        client.connectTimeout = 30000 // 30 seconds
        client.responseTimeout = 60000 // 60 seconds
        // Pretend we are cURL, so that Facebook does not redirect us to facebook.com/unsupportedbrowser
        client.setUserAgent("curl/7.55.1")
        return client.get(birthdayICalUri, handler)
    }
}
